File: comp/ADT/Java/ADT.java
// Example of sum types
//   Very easy - goes back to its C heritage
class Animal {
    int numberLegs;
    String name;
}

class Eight {
    boolean ones;
    boolean twos;
    boolean fours;
}

// Example of product types
//   Rather forced - In Java data types like to be
//                   either one thing or another.
class IntOrString {
    private IntOrString() {}
    class AnInt extends IntOrString {
        int i;
        AnInt(int i) { this.i = i; }
        Boolean isInt() { return true; }
        Boolean isString() { return false; }
    }
    class AnString extends IntOrString {
        String s;
        AnString(String s) { this.s = s; }
        Boolean isInt() { return false; }
        Boolean isString() { return true; }
    }

}

public class ADT {
    public static void main(String[] args) {
        System.out.println("See if things even compile");
    }
}

File: comp/ADT/Scala/ADT.scala
package comp.adt

// Example of product types

/* This one essentially the same as a Tuple3.
 * It has 8 values if A = B = C = Boolean
 */
case class Eight[A,B,C](a: A, b: B, c: C)

// Examples of sum types

/* This one has 6 distinct values */
sealed trait Six
case class Bool1(n: Boolean) extends Six
case class Bool2(n: Boolean) extends Six
case class Bool3(n: Boolean) extends Six

sealed trait IntOrString
case class AnInt(i: Int) extends IntOrString
case class AnString(s: String) extends IntOrString

// Test the above.
object ADT {

  def main(args: Array[String]) = {
    println("See if it compiles")
  }
}
